### ISSUE TITLE
[rllib] Add support for auxiliary/internal rewards

### ISSUE THREAD

**Issue by jon-chuang** [ray-project/ray#5470]
### Describe the problem
We should add abstractions that support a wide variety of internal rewards, which in general require only trajectory data to compute, for example:

Exploration:
RND, ICM and Reachability Memory (https://arxiv.org/pdf/1810.02274.pdf)
Go-Explore (https://eng.uber.com/go-explore/)

Auxiliary:
Pixel Control

Fear - staying in safe expert distribution:
GAIL (observations and actions)

I believe this can be accomplished solely with `postprocessing_fn`.

In addition, of course, we should implement the above algorithms.

It is possible that due to modularity, code written in tensorflow, say for the RL model, could be interacting with pytorch code implementing an intrinsic reward module, since they are often trained to optimise different targets :) However, I'm not familiar enough with the codebase to say how likely this is.

I would also like to see how RND/Go-Explore stacks up against Reachability in Montezuma's Revenge, since no such comparison has been made so far.
